shared-libs/cht-datasource/src/input.ts

```diff
@@ -0,0 +1,270 @@
+/** @internal */
+const checkContactInputFields = (data: unknown): data is Record<string, unknown> => {
+  return isRecord(data) &&
+    hasFields(data, [
+      { name: 'type', type: 'string', ensureTruthyValue: true },
+      { name: 'name', type: 'string', ensureTruthyValue: true }
+    ]) &&
+    (!('reported_date' in data) || isValidReportedDate(data.reported_date));
+};
+
+/**
+ * An input for a report
+ */
+export type ReportInput = Readonly<{
+  type: string,
+  form: string,
+  reported_date?: string | number,
+  _id?: string,
+  _rev?: string,
+  contact: string
+}>;
+
+/**
+ * Builds an input object for creation and update of a report with
+ * the given fields.
+ * @param data object containing the fields for a report
+ * @returns the report input
+ * @throws Error if data is not an object
+ * @throws Error if type is not provided or is empty
+ * @throws Error if form is not provided or is empty
+ * @throws Error if contact is not provided or is empty
+ * @throws Error if reported_date is not in a valid format.
+ * Valid formats are 'YYYY-MM-DDTHH:mm:ssZ', 'YYYY-MM-DDTHH:mm:ss.SSSZ', or <unix epoch>.
+ */
+export const validateReportInput = (data: unknown): ReportInput => {
+  if (!isRecord(data)) {
+    throw new InvalidArgumentError('Invalid "data": expected an object.');
+  }
+
+  const input = { ...data };
+  insertReportedDateIfMissing(input);
+  if (!isValidReportedDate(input.reported_date)) {
+    throw new InvalidArgumentError(
+      'Invalid reported_date. Expected format to be ' +
+      '\'YYYY-MM-DDTHH:mm:ssZ\', \'YYYY-MM-DDTHH:mm:ss.SSSZ\', or a Unix epoch.'
+    );
+  }
+  input.reported_date = convertToUnixTimestamp(input.reported_date as string | number);
+  if (!hasField(input, { name: 'contact', type: 'string', ensureTruthyValue: true })) {
+    throw new InvalidArgumentError(`Missing or empty required field (contact) in [${JSON.stringify(data)}].`);
+  }
+  if (!isReportInput(input)) {
+    throw new InvalidArgumentError(`Missing or empty required field (form) in [${JSON.stringify(data)}].`);
+  }
+  return input;
+};
+
+/**
+ * Returns `true` if the given input is a {@link ReportInput} otherwise `false`.
+ * @param input the input to check
+ * @returns `true` if the given type is a {@link ReportInput}, otherwise `false`.
+ */
+export const isReportInput = (input: unknown): input is ReportInput => {
+  if (isRecord(input) &&
+    hasFields(input, [
+      { name: 'form', type: 'string', ensureTruthyValue: true },
+      { name: 'contact', type: 'string', ensureTruthyValue: true }
+    ])
+  ) {
+    if ('reported_date' in input && !isValidReportedDate(input.reported_date)) {
+      return false;
+    }
+    return true;
+  }
+  return false;
+};
+
+
+/** @internal */
+const isValidReportedDate = (value: unknown): boolean => {
+  if (typeof value === 'number') {
+    return Number.isFinite(value);
+  }
+
+  if (typeof value === 'string') {
+    const isoRegex = /^(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d{3})?Z)$/;
```

Comment:
Minor, but would be nice to just have this pattern as a file-level constant.  Maybe `ISO_8601_DATE_PATTERN`?